pandas/core/reshape/reshape.py

```diff
@@ -1096,6 +1100,162 @@ def get_empty_frame(data) -> DataFrame:
         return DataFrame(dummy_mat, index=index, columns=dummy_cols)
 
 
+def from_dummies(
+    data: DataFrame,
+    sep: None | str = None,
+    implied_category: None | Hashable | dict[str, Hashable] = None,
+) -> DataFrame:
+    """
+    Create a categorical `DataFrame` from a `DataFrame` of dummy variables.
+
+    Inverts the operation performed by 'get_dummies'.
+
+    Parameters
+    ----------
+    data : DataFrame
+        Data which contains dummy-coded variables.
+    sep : str, default None
+        Separator used in the column names of the dummy categories they are
+        character indicating the separation of the categorical names from the prefixes.
+        For example, if your column names are 'prefix_A' and 'prefix_B',
+        you can strip the underscore by specifying sep='_'.
+    implied_category : None, Hashable or dict of Hashables, default None
+        The implied category the dummy takes when all values are zero.
+        Can be a a single value for all variables or a dict directly mapping the
+        implied categories to a prefix of a variable.
+
+    Returns
+    -------
+    DataFrame
+        Categorical data decoded from the dummy input-data.
+
+    Notes
+    -----
+    The columns of the passed dummy data should only include 1's and 0's,
+    or boolean values.
+
+    See Also
+    --------
+    get_dummies : Convert `Series` or `DataFrame` to dummy codes.
+
+    Examples
+    --------
+    >>> df = pd.DataFrame({"a": [1, 0, 0, 1], "b": [0, 1, 0, 0],
+    ...                    "c": [0, 0, 1, 0]})
+
+    >>> pd.from_dummies(df)
+    0     a
+    1     b
+    2     c
+    3     a
+
+    >>> df = pd.DataFrame({"col1_a": [1, 0, 1], "col1_b": [0, 1, 0],
+    ...                    "col2_a": [0, 1, 0], "col2_b": [1, 0, 0],
+    ...                    "col2_c": [0, 0, 1]})
+
+    >>> pd.from_dummies(df, sep="_")
+        col1    col2
+    0    a       b
+    1    b       a
+    2    a       c
+
+    >>> df = pd.DataFrame({"col1_a": [1, 0, 0], "col1_b": [0, 1, 0],
+    ...                    "col2_a": [0, 1, 0], "col2_b": [1, 0, 0],
+    ...                    "col2_c": [0, 0, 0]})
+
+    >>> pd.from_dummies(df, sep="_", implied_category={"col1": "d", "col2": "e"})
+        col1    col2
+    0    a       b
+    1    b       a
+    2    d       e
+    """
+    from pandas.core.reshape.concat import concat
+
+    if data.isna().any().any():
+        raise ValueError(
+            f"Dummy DataFrame contains NA value in column: "
+            f"'{data.isna().any().idxmax()}'"
+        )
```

Comment:
`mypy` is currently unhappy with:
```
pandas/core/reshape/reshape.py:1174: error: Item "bool" of "Union[Series, bool]" has no attribute "any"  [union-attr]
pandas/core/reshape/reshape.py:1176: error: Item "bool" of "Union[Series, bool]" has no attribute "idxmax"  [union-attr]
```
as `data` is always of type `DataFrame` -> `data.isna().any()` always returns a `Series` if I my understanding is correct. 
I will look into it.

Reply:
> data.isna().any() always returns a Series if I my understanding is correct.

yes, but `mypy` doesn't know that 😆  mypy only knows that the function signatures tell it, and `any` is typed to return `Series | bool` in `pandas/core/generic.py`. This could be solved by defining `any` in `pandas/core/frame.py` and `pandas/core/series.py` such that the former returns `Series` and the latter returns `bool`

Reply:
If you wanted to address that in a separate PR, that'd be great - for now, in this one, I'd just `# type: ignore` it, and put the error message on top as a comment (see other `# type: ignore` comments for examples)

Reply:
Okay, will quick-fix the issues by ignoring them and take a look at the requried typed returns.